File: src/morie/fn/aitpow.py
# morie.fn -- function file (rootcoder007/morie)
"""Powering: scalar multiplication on the simplex."""

from . import _tail1core as C

from ._richresult import RichResult

__all__ = ["powering", "aitchison_powering"]


def powering(a, x, total=1.0):
    """Power a composition by a real scalar.

    Powering is scalar multiplication in the Aitchison vector space:
    together with perturbation it makes the simplex a real vector
    space, so "twice the effect" has an unambiguous meaning.  a > 1
    pushes a composition towards its dominant part, 0 < a < 1 pulls it
    towards the barycentre, and a < 0 reverses it.

    Formula: a (.) x = C( x_1^a, ..., x_D^a )

    Parameters
    ----------
    a : float
        Scalar.
    x : array-like
        Strictly positive vector of parts.
    total : float
        Constant the result sums to.

    Returns
    -------
    RichResult
        ``composition``, ``a``, ``total``, ``D``.

    References
    ----------
    Aitchison (1986), The Statistical Analysis of Compositional Data,
    Chapter 2.  Verified against the reference implementation in the
    CRAN package ``compositions`` 2.0-9, whose power operator on an
    ``acomp`` closes the elementwise power.
    """
    x = C.vec(x)
    if any(v <= 0 for v in x):
        raise ValueError("compositions must be strictly positive")
    a = float(a)
    p = [v ** a for v in x]
    s = sum(p)
    k = float(total)
    return RichResult(payload={
        "composition": [k * v / s for v in p], "a": a, "total": k,
        "D": len(x), "method": "Powering on the simplex"})


aitchison_powering = powering


def cheatsheet():
    return "aitpow: a (.) x = C(x^a)"

File: r-package/morie/R/aitpow.R
# SPDX-License-Identifier: AGPL-3.0-or-later
#' Powering, the scalar multiplication of the simplex.
#'
#' Formula: a (.) x = C( x_1^a, x_2^a, ..., x_D^a )
#'
#' @param x Composition with strictly positive parts.
#' @param a Real scalar.
#' @param total Constant kappa the closure sums to.
#'
#' @return List with ``composition``, ``alpha``, ``total``, ``D``.
#' @references Aitchison, J. (1986), The Statistical Analysis of Compositional Data, Chapman and Hall, is this shelf's primary book and is NOT in the reference library, so it could not be read.  The definitions below were taken instead from Mateu-Figueras, G., Pawlowsky-Glahn, V. and Egozcue, J. J., The normal distribution in some constrained sample spaces, arXiv:0802.2643 (published as SORT 37(1):29-56, 2013), Sect. 4.1, which prints them with equation numbers and attributes them to Aitchison (1982, 1986); that paper was FETCHED and is archived in the reference library with a row in EXTERNAL_SOURCES.md.  Powering is the outer product of the simplex: raise every part to the power a and close.  It plays the role that multiplication by a scalar plays in real space, and satisfies d_a(a(.)x, a(.)x*) = |a| d_a(x, x*).
#' @export
Comppower <- function(x, a, total = 1) {
  x <- .t1_vec(x)
  if (length(x) == 0L) stop("x must be non-empty")
  if (any(x <= 0)) stop("compositions must be strictly positive")
  a <- as.numeric(a); lg <- a * log(x)
  e <- exp(lg - max(lg)); s <- sum(e); k <- as.numeric(total)
  .t1_result(composition = k * e / s, alpha = a, total = k, D = length(x),
             method = "Powering on the simplex")
}
